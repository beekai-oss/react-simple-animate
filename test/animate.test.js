import React from 'react';
import Animate from '../src/animate';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';

jest.useFakeTimers();

const props = {
  startAnimation: true,
  startStyle: {},
  endStyle: {},
  durationSeconds: 1,
  className: 'test',
};

const endStyle = { display: 'block' };

describe('Animate', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Animate {...props}>test</Animate>);

    expect(tree).toMatchSnapshot();
  });

  it('should render custome tag correctly', () => {
    const tree = renderer.create(<Animate {...{ ...props, tag: 'l' }} />);

    expect(tree).toMatchSnapshot();
  });

  describe('when animation is ready and delay have been set', () => {
    it('should trigger animation delay ended', () => {
      const tree = mount(
        <Animate {...{ ...props, delaySeconds: 1 }}>test</Animate>,
      );

      jest.runAllTimers();
      expect(tree.state().delayWillEnd).toEqual(true);
    });
  });

  describe('when animate end style is defined and ready to animate', () => {
    it('should trigger animation about to end state', () => {
      const tree = mount(
        <Animate
          {...{
            ...props,
            startAnimation: false,
            onCompleteStyle: endStyle,
          }}
        >
          test
        </Animate>,
      );

      tree.setProps({
        startAnimation: true,
      });

      jest.runAllTimers();
      expect(tree.state().animationWillEnd).toEqual(true);
    });
  });

  it('should update component when animation is ready', () => {
    const tree = shallow(
      <Animate {...{ ...props, startAnimation: false, endStyle: endStyle }}>
        test
      </Animate>,
    );

    const state = {
      animationWillEnd: false,
      delayWillEnd: false,
    };

    expect(
      tree.instance().shouldComponentUpdate(
        {
          startAnimation: true,
        },
        state,
      ),
    ).toEqual(true);
  });

  it('should not update component when animation already played', () => {
    const tree = shallow(
      <Animate {...{ ...props, startAnimation: false, endStyle: endStyle }}>
        test
      </Animate>,
    );

    const state = {
      animationWillEnd: false,
      delayWillEnd: false,
    };

    const nextProps = {
      startAnimation: false,
      children: 'test',
      startStyle: {},
      endStyle,
    };

    expect(
      tree.instance().shouldComponentUpdate(
        {
          startAnimation: true,
        },
        state,
      ),
    ).toEqual(true);

    expect(tree.instance().shouldComponentUpdate(nextProps, state)).toEqual(
      false,
    );
  });

  it('should update component when animation is about to end or delay animation about to end ', () => {
    const tree = shallow(
      <Animate
        {...{ ...props, startAnimation: false, endStyle: { display: 'block' } }}
      >
        test
      </Animate>,
    );

    let state = {
      animationWillEnd: true,
      delayWillEnd: false,
    };

    const nextProps = {
      startAnimation: false,
    };

    expect(tree.instance().shouldComponentUpdate(nextProps, state)).toEqual(
      true,
    );

    state = {
      animationWillEnd: false,
      delayWillEnd: true,
    };

    expect(tree.instance().shouldComponentUpdate(nextProps, state)).toEqual(
      true,
    );
  });

  describe('When animation is about to end', () => {
    it('should update the style', () => {
      const onCompleteStyle = {
        display: 'block',
      };
      const tree = shallow(
        <Animate
          {...{
            ...props,
            startAnimation: false,
            onCompleteStyle,
          }}
        >
          test
        </Animate>,
      );

      tree.setState({
        animationWillEnd: true,
      });

      expect(tree.find('div').props().style).toEqual({
        ...onCompleteStyle,
        transition: null,
      });
    });
  });

  describe('when animation about to end', () => {
    it('should update the style', () => {
      const endStyle = {
        display: 'block',
      };
      const tree = shallow(
        <Animate
          {...{
            ...props,
            startAnimation: true,
            delaySeconds: 0,
            endStyle,
          }}
        >
          test
        </Animate>,
      );

      expect(tree.find('div').props().style).toEqual({
        ...endStyle,
        transition: '1s all linear',
      });
    });
  });

  describe('when delay animation about to end', () => {
    it('should update the style', () => {
      const endStyle = {
        display: 'block',
      };
      const tree = shallow(
        <Animate
          {...{
            ...props,
            delaySeconds: 1,
            endStyle,
          }}
        >
          test
        </Animate>,
      );

      tree.setState({
        delayWillEnd: true,
      });

      expect(tree.find('div').props().style).toEqual({
        ...endStyle,
        transition: '1s all linear',
      });
    });
  });

  describe('when is animate prop value get changed', () => {
    it('should trigger state reset', () => {
      const tree = shallow(
        <Animate
          {...{
            ...props,
          }}
        >
          test
        </Animate>,
      );

      tree.setState({
        animationWillEnd: true,
        delayWillEnd: true,
      });

      tree.setProps({
        startAnimation: false,
      });

      expect(tree.state()).toEqual({
        animationWillEnd: false,
        delayWillEnd: false,
      });
    });
  });

  describe('when force update passed as prop', () => {
    it('should re-render the component', () => {
      const tree = shallow(
        <Animate
          {...{
            ...props,
          }}
        >
          test
        </Animate>,
      );

      const nextProps = {
        ...props,
        forceUpdate: true,
      };

      const nextState = { animationWillEnd: false, delayWillEnd: false };

      expect(
        tree.instance().shouldComponentUpdate(nextProps, nextState),
      ).toEqual(true);
    });
  });

  describe('when children component updated', () => {
    it('should trigger re-render', () => {
      const tree = shallow(
        <Animate
          {...{
            ...props,
          }}
        >
          <div>test</div>
        </Animate>,
      );

      const nextProps = {
        ...props,
        children: (
          <div>
            <span>test</span>
          </div>
        ),
      };

      const nextState = { animationWillEnd: false, delayWillEnd: false };

      expect(
        tree.instance().shouldComponentUpdate(nextProps, nextState),
      ).toEqual(true);
    });
  });

  describe('when transition have been passed down', () => {
    it('should overwrite the transition style', () => {
      const transition = '1s opacity';
      const tree = shallow(
        <Animate
          {...{
            ...props,
            transition,
          }}
        />,
      );

      expect(tree.find('div').props().style).toEqual({ transition });
    });
  });
});
