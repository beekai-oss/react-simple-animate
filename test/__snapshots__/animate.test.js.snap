// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animate should render correctly 1`] = `
<div
  className="test"
  style={
    Object {
      "transition": "1s all linear",
    }
  }
>
  test
</div>
`;

exports[`Animate should render correctly for children 1`] = `
<div
  className="test"
  style={
    Object {
      "transition": "1s all linear",
    }
  }
>
  <div>
    test
    1
  </div>
  <div>
    test
    2
  </div>
</div>
`;

exports[`Animate should render correctly with render props 1`] = `
<div
  className="test"
  style={
    Object {
      "transition": "1s all linear",
    }
  }
>
  what
</div>
`;

exports[`Animate should render custome tag correctly 1`] = `
<l
  className="test"
  style={
    Object {
      "transition": "1s all linear",
    }
  }
/>
`;

exports[`Animate when animate on add or remove enabled should render correctly for Animate children 1`] = `
<div
  className="test"
  style={
    Object {
      "transition": "1s all linear",
    }
  }
>
  <div
    className={undefined}
    style={
      Object {
        "transition": "0.3s all linear",
      }
    }
  >
    <div>
      test
      1
    </div>
  </div>
  <div
    className={undefined}
    style={
      Object {
        "transition": "0.3s all linear",
      }
    }
  >
    <div>
      test
      2
    </div>
  </div>
</div>
`;

exports[`Animate when animate on add or remove enabled should render correctly for children 1`] = `
<div
  className="test"
  style={
    Object {
      "transition": "1s all linear",
    }
  }
>
  <div
    style={
      Object {
        "transition": "1s all linear",
      }
    }
  >
    test
    1
  </div>
  <div
    style={
      Object {
        "transition": "1s all linear",
      }
    }
  >
    test
    2
  </div>
</div>
`;
